contracts/BullaFinance.sol

```diff
@@ -0,0 +1,131 @@
+    ///         P4. `terms.termCompletionTimestamp < type(uint40).max`
+    ///         P5. `terms.minDownPaymentBPS < type(uint24).max`
+    ///         P6. `terms.interestBPS < type(uint24).max`
+    ///         P7. `terms.termCompletionTimestamp < type(uint40).max`
+    ///         P8. `terms.termCompletionTimestamp > block.timestamp` TODO: necessary?
+    function createInvoiceWithFinanceOffer(BullaBanker.ClaimParams memory claim, FinanceTerms memory terms)
+        public
+        virtual
+        returns (uint256 claimId);
+
+    /// @param claimId the id of the underlying claim
+    /// @param terms financing terms
+    /// @notice SPEC:
+    ///     Allows a creditor to offer financing on an existing pending claim OR update previously offerred financing terms // TODO: should this function be used to rescind a financing offer?
+    ///     This function will:
+    ///         RES1. Overwrite the `terms` as indexed by the specified `claimId`
+    ///         RES2. Emit a FinancingOffered event
+    ///     Given the following:
+    ///         P1. `claim.status == ClaimStatus.Pending`
+    ///         P2. `msg.sender == claim.creditor`
+    ///         P3. if terms[claimId].termCompletionTimestamp == 0 (implying new terms on an existing claim) ensure msg.value == fee
+    ///         P4. `(terms.minDownPaymentBPS * claim.claimAmount / 10_000) > 0`
+    ///         P5. `terms.minDownPaymentBPS < type(uint24).max`
+    ///         P6. `terms.interestBPS < type(uint24).max`
+    ///         P7. `terms.termCompletionTimestamp < type(uint40).max`
+    ///         P8. `terms.termCompletionTimestamp > block.timestamp` TODO: necessary?
+    function offerFinancing(uint256 claimId, FinanceTerms memory terms) public virtual;
+
+    /// @param claimId the id of the underlying claim
+    /// @notice SPEC:
+    ///     Allows a creditor to reclaim feeAmount of tokens if the underlying claim is no longer pending
+    ///     This function will:
+    ///         RES1. delete `financeTerms[claimId]`
+    ///         RES2. transfer the creditor `fee` amount of tokens
+    ///         RES3. Emit a FeeReclaimed event with the underlying claimId
+    ///     Given the following:
+    ///         P1. `claim.status != ClaimStatus.Pending`
+    function reclaimFee(uint256 claimId) public virtual;
+
+    //
+    //// DEBTOR FUNCTIONS ////
+    //
+
+    /// @param claimId id of the originating claim
+    /// @param downPayment the amount the debtor wishes to contribute
+    /// @notice SPEC:
+    ///     Allows a debtor to accept a creditor's financing offer and begin payment
+    ///     This function will:
+    ///         RES1. load the previous claim details and create a new bulla claim specifying `claimAmount` as `originatingClaimAmount + (originatingClaimAmount * terms.interestBPS / 10_000)` and `dueBy` as `term.termCompletionTimestamp`
+    ///         RES2. deletes the `financeTerms` TODO: does this have any potential negative side-effects / drawbacks?
```

Comment:
not sure we have to delete. Could be useful to keep this data here. it would only be available offchain in that case.

Reply:
Good point